aiida/tools/ipython/ipython_magics.py

```diff
@@ -66,14 +66,41 @@ class AiiDALoaderMagics(magic.Magics):
     @magic.needs_local_scope
     @magic.line_magic
     def verdi(self, line='', local_ns=None):  # pylint: disable=no-self-use,unused-argument
-        """Run the AiiDA command line tool, using the currently loaded configuration and profile."""
+        """Run the AiiDA command line tool, using the currently loaded configuration and profile.
+
+        Invoking ``verdi`` normally through the command line follows a different code path, compared to calling it
+        directly from within an active Python interpreter. Some of those code paths we actually need here, and others
+        can actually cause problems:
+
+        * The ``VerdiCommandGroup`` group ensures that the context ``obj`` is set with the loaded ``Config`` and
+          ``Profile``, but this is not done in this manual call, so we have to built it ourselves and pass it in with
+          the ``obj`` keyword.
+        * We cannot call the ``aiida.cmdline.commands.cmd_verdi.verdi`` command directly, as that will invoke the
+          ``aiida.cmdline.parameters.types.profile.ProfileParamType`` parameter used for the ``-p/--profile`` option
+          that is defined on the ``verdi`` command. This will attempt to load the default profile, but here a profile
+          has actually already been loaded. In principle, reloading a profile should not be a problem, but this magic
+          is often used in demo notebooks where a temporary profile was created and loaded, which is not properly added
+          to the config file (since it is temporary) and so loading the profile would fail. The solution is to not call
+          the top-level ``verdi`` command, but the subcommand, which is the first parameter in the command line
+          arguments defined by the ``line`` argument.
+
+        """
+        from click import Context
+
         from aiida.cmdline.commands.cmd_verdi import verdi
         from aiida.common import AttributeDict
         from aiida.manage import get_config, get_profile
-        config = get_config()
-        profile = get_profile()
-        obj = AttributeDict({'config': config, 'profile': profile})
-        return verdi(shlex.split(line), prog_name='%verdi', obj=obj, standalone_mode=False)  # pylint: disable=too-many-function-args,unexpected-keyword-arg
+
+        cmdline_arguments = shlex.split(line)
+        context = Context(verdi)
```

Comment:
Can you give me an example of what this context will be. Is this the profile name passed to the verdi command when `-p` is used?

Reply:
You are right to point this out, if the user actually calls `%verdi -p some-profile node show` that will fail. But even if we were to parse it correctly, it would still fail if the specified profile is other than the currently loaded one. I think this behavior should not be supported. The only reason for the `%verdi` magic, as I understand it, is to be able to use a verdi command for the currently loaded profile in a notebook.

Reply:
I see. Then I think we can raise an exception explicitly if the user pass `-p`.

Reply:
Need to be careful and also check for `--profile`. But yeah, I can add that.

Reply:
I have added a check and raise if the profile parameter is specified.